fv3core/utils/future_stencil.py

```diff
@@ -0,0 +1,307 @@
+    wrapper: Optional[Callable] = None,
+    rebuild: bool = False,
+    **kwargs: Any,
+):
+    """
+    Create a future stencil object with deferred building in a distributed context
+
+    Parameters
+    ----------
+        backend : `str`
+            Name of the implementation backend.
+
+        definition : `None` when used as a decorator, otherwise a `function` or a
+                     `:class:`gt4py.StencilObject`
+            Function object defining the stencil.
+
+        externals: `dict`, optional
+            Specify values for otherwise unbound symbols.
+
+        rebuild : `bool`, optional
+            Force rebuild of the :class:`gt4py.StencilObject` even if it is
+            found in the cache. (`False` by default).
+
+        **kwargs: `dict`, optional
+            Extra backend-specific options. Check the specific backend
+            documentation for further information.
+
+    Returns
+    -------
+        :class:`FutureStencil`
+            Wrapper around an instance of the dynamically-generated subclass
+            of :class:`gt4py.StencilObject`.
+    """
+
+    def _decorator(func):
+        device_sync = kwargs.pop("device_sync", None)
+        backend_opts = {"device_sync": device_sync} if device_sync else {}
+        builder = (
+            StencilBuilder(func)
+            .with_backend(backend)
+            .with_externals(externals or {})
+            .with_options(
+                name=func.__name__,
+                module=func.__module__,
+                rebuild=rebuild,
+                backend_opts=backend_opts,
+                **kwargs,
+            )
+        )
+        return FutureStencil(builder, wrapper)
+
+    if definition is None:
+        return _decorator
+    return _decorator(definition)
+
+
+class FutureStencil:
+    """
+    A stencil object that is compiled by this node or another in a distributed context.
+    """
+
+    _id_table = StencilTable()
+
+    def __init__(
+        self,
+        builder: Optional["StencilBuilder"] = None,
+        wrapper: Optional[Callable] = None,
+    ):
+        self._builder: Optional["StencilBuilder"] = builder
+        self._stencil_object: Optional[StencilObject] = None
+        self._sleep_time: float = 0.05
+        self._timeout: float = 180.0
+        self._node_id = MPI.COMM_WORLD.Get_rank() if MPI else 0
+        self._wrapper = wrapper
+
+    @classmethod
+    def clear(cls):
+        cls._id_table.clear()
+
+    @property
+    def cache_info_path(self) -> str:
+        return self._builder.caching.cache_info_path.stem
+
+    @property
+    def stencil_object(self) -> StencilObject:
+        if self._stencil_object is None:
+            self._wait_for_stencil()
+        return self._stencil_object
+
+    @property
+    def field_info(self) -> Dict[str, FieldInfo]:
+        return self.stencil_object.field_info
+
+    def _delay(self) -> float:
+        delay_time = self._sleep_time * float(self._node_id)
```

Comment:
```suggestion
        delay_time = self._sleep_time + float(self._node_id)
```
With the current `_sleep_time`, node 1000 would wait 50s

Reply:
Actually that won't work either! Should be doing `_sleep_time*float(self._node_id) / 10` or something

Reply:
Great idea.